model/modeldecoder/v2/decoder.go

```diff
@@ -0,0 +1,225 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+func init() {
+	metadataRootPool.New = func() interface{} {
+		return &metadataRoot{}
+	}
+	// metadataNoKeyPool.New = func() interface{} {
+	// 	return &metadataNoKey{}
+	// }
+}
+
+var metadataRootPool sync.Pool
```

Comment:
```suggestion
var metadataRootPool = sync.Pool{
	New: func() interface{} {
		return &metadataRoot{}
	},
}
```

I think I previously suggested using `init()` instead of assigning the result of calling a function literal to a package-level var. Not sure if you that's why you used `init()` here. In this case we're just assigning a function literal (not calling it), so I'd go for this simpler approach.